### DISCUSSION TITLE
RingParameters class

### DISCUSSION THREAD

**Discussion by simoneliuzzo** [atcollab/at#474] [Python AT]
when using a RingParameters object it seems not possible to use the standard  python dictionary like indexing

<img width="1105" alt="Screenshot 2022-08-31 at 13 08 38" src="https://user-images.githubusercontent.com/30342742/187665264-792423b8-faea-4999-806b-b99b5f31d225.png">

I personally find this a limitation, but may be it is intentional.

**Comment by simoneliuzzo** [comment]
I found a solution to restore my wished way to access data: `datae[0].__dict__['emittances']`

**Comment by lfarv** [comment]
Hello @simoneliuzzo,

You are trying to add complexity to simple things !
`RingParameters` are a class objects, not python `dict`. So you access their attributes, as for any python object, with the "dot" notation (like for a Matlab `struct` or class object):
```python
>>> params = ring.radiation_parameters()
>>> params.emittances
array([1.31487971e-10,            nan,            nan])
```

Using a class object rather than a `dict` allows to implement custom methods, for instance pretty printing:
```python
>>> print(params)
              Frac. tunes: [0.2099983  0.34001317 0.00349013]
                    Tunes: [76.2099983  27.34001317]
           Chromaticities: [5.73409894 3.91761206]
 Momentum compact. factor: 8.506669e-05
              Slip factor: -8.505944e-05
                   Energy: 6.000000e+09 eV
       Energy loss / turn: 2.526189e+06 eV
 Radiation integrals - I1: 0.07179435013387388 m
                       I2: 0.13844595446798158 m^-1
                       I3: 0.003357584058614851 m^-2
                       I4: -0.07375725030666251 m^-1
                       I5: 5.281495714523264e-07 m^-1
          Mode emittances: [1.31487971e-10            nan            nan]
Damping partition numbers: [1.53275121 1.         1.46724879]
            Damping times: [0.00872477 0.0133729  0.00911427] s
            Energy spread: 0.000934463
             Bunch length: 0.0030591 m
         Cavities voltage: 6000000.0 V
        Synchrotron phase: 2.70701 rd
    Synchrotron frequency: 1239.74 Hz
```

With your syntax, you are extracting a `dict` for the object. Be careful, because this only extract a subset of an object's attributes. You are missing the object's properties and the `__slot__` attributes (there are no such things in `RingParameters`)

**Comment by simoneliuzzo** [comment]
I am saving RingParameters objects for many lattice and then reloading them to make histograms of each of the parameters inside RingParameters. I have a single function to do the histogram and I wanted to pass the name (and index) of the parameter to plot. The dot notation does not allow me to pass the name of the wished property as a variable, while `__dict__` does. It must be a very specific case of use.

**Comment by lfarv** [comment]
No problem: if the attribute name is in a variable, use the [`getattr`](https://docs.python.org/fr/3/library/functions.html#getattr) function !
